src/design_notes/copy_ergonomics.md

```diff
@@ -0,0 +1,105 @@
+# Copy type ergonomics
+
+## Background
+
+There are a number of pain points with `Copy` types that the lang team is
+interested in exploring, though active experimentation is not currently ongoing.
+
+Some key problems are:
+
+## `Copy` cannot be implemented with non-`Copy` members
+
+There are standard library types where the lack of a `Copy` impl is an
+active pain point, e.g., [`MaybeUninit`](https://github.com/rust-lang/rust/issues/62835)
+and [`UnsafeCell`](https://github.com/rust-lang/rust/issues/25053)
+
+### History
+
+ * `unsafe impl Copy for T` which avoids the requirement that T is recursively
+   Copy, but is obviously unsafe.
+    * https://github.com/rust-lang/rust/issues/25053#issuecomment-218610508
+ * `Copy` is dangerous on types like `UnsafeCell` where `&UnsafeCell<T>`
+   otherwise would not permit access to `T` in [safe
+   code](https://github.com/rust-lang/rust/issues/25053#issuecomment-98447164).
+
+## `Copy` types can be (unintentionally) copied
+
+Even if a type is Copy (e.g., `[u8; 1024]`) it may not be a good idea to make
+use of that in practice, since copying large amounts of data is slow. This is
+primarily a performance concern, so the problem is usually that these copies are
+easy to miss. It is also true that linting on these cases is really difficult,
+as depending on the problem, it can be quite reasonable to produce
+megabyte-large copies once per program, but in a web server doing so on each
+request may not be a good idea.
+
+Implementations of `Copy` on closures and arrays are the prime example of Rust
+currently being overeager with the defaults in some contexts.
+
+This also comes up with `Copy` impls on `Range`, which would generally be
+desirable but is error-prone given the `Iterator/IntoIterator` impls on ranges.
+
+The example here does not compile today (since Range is not Copy), but would be
+unintuitive if it did.
+
+```rust,compile_fail
+let mut x = 0..10;
+let mut c = move || x.next();
+println!("{:?}", x.next()); // prints 0
+println!("{:?}", c()); // prints 0, because the captured x is implicitly copied.
+```
```

Comment:
Why is this unintuitive? `move` explicitly asks to *not* capture `x` by reference.